src/Network.hs

```diff
@@ -31,34 +31,47 @@ data ForwardResult a = ForwardResult {
+                M.Matrix a                                                  -- ^ Input of the layer
+                -> A.Activation                                             -- ^ Activation function
+                -> M.Matrix a                                               -- ^ Weights of current layer
+                -> M.Matrix a                                               -- ^ Ouput of the previous layer
+                -> M.Matrix a                                               -- ^ Current propagation
+                -> Either String ((M.Matrix a, M.Matrix a), M.Matrix a)     -- ^ Gradients of the current layer, with the propagation vector
 backwardStep input activation weights output propagation = do
     jacobian <- A.backward activation input
-    nextPropagation <- M.multiplyMatrices propagation jacobian >>= flip M.multiplyMatrices weights
-    nextGradient <- M.transpose <$> (M.multiplyMatrices propagation jacobian >>= M.multiplyMatrices output)
-    return (nextGradient, nextPropagation)
+    let buffer = M.multiplyMatrices propagation jacobian
+    nextPropagation <- buffer >>= flip M.multiplyMatrices weights
+    nextWeightsGradient <- M.transpose <$> (buffer >>= M.multiplyMatrices output)
+    nextBiasesGradient <- M.transpose <$> buffer
+    return ((nextBiasesGradient, nextWeightsGradient), nextPropagation)
 
 forwardStep :: RealFloat a =>
-    M.Matrix a                                   -- ^ Input of the network
-    -> Network a                                 -- ^ Current network
-    -> Either String [(M.Matrix a, M.Matrix a)]  -- ^ Inputs and outputs of the current layer.
-forwardStep input (Network [] []) = Right []
-forwardStep input (Network (activation:activations) (weight:weights)) = do
-    activationInput <- M.multiplyMatrices weight input
-    output <- A.forward activation activationInput
-    ((activationInput, output) :) <$> forwardStep output (Network activations weights)
+               M.Matrix a                                   -- ^ Input of the network
+               -> Network a                                 -- ^ Current network
+               -> Either String [(M.Matrix a, M.Matrix a)]  -- ^ Inputs and outputs of the current layer.
+forwardStep input (Network [] [] []) = Right []
+forwardStep input (Network (activation:activations) (bias:biases) (weight:weights)) = do
+    activationInput <- M.multiplyMatrices weight input >>= M.addMatrices bias
+    activationOutput <- A.forward activation activationInput
+    ((activationInput, activationOutput) :) <$> forwardStep activationOutput (Network activations biases weights)
+
+randomStep :: [Int] -> SR.StdGen -> Either String ([M.Matrix Double], [M.Matrix Double])
+randomStep [layer] _ = Right ([], [])
+randomStep (firstLayer : secondLayer : nextLayers) generator = do
+    let (firstGenerator, secondGenerator) = SR.split generator
+    let (_, thirdGenerator) = SR.split secondGenerator
```

Comment:
This can be made 15% more fun and 25% less readable with:
`let (firstGenerator, (secondGenerator, thirdGenerator)) = SR.split <$> SR.split generator `